triplea-game/triplea
# Decouple headless game server from assets folder

## Overview

Decouples the headless game server from the _assets_ folder so that the headless game server artifact does not have to include this content.  Removal of the _assets_ folder from the headless game server artifact reduces its size from 17.1 MiB to 9.84 MiB (-42%).

This PR also finally gets rid of the smell of having to run the `game-headed` and `game-headless` projects from the `game-core` folder in a development environment.

## Functional Changes

* Moved the _assets_ and _dice_servers_ folders to the `game-headed` project; moved the _.triplea-root_ marker file to both the `game-headed` and `game-headless` projects.
* Updated the Eclipse launch configurations, the Gradle `run` tasks, and the Gradle `integTest` task to run from the corresponding project directory rather than the `game-core` project directory.
* Introduced a new service (`ApplicationContext`) so that code in `game-core` can get access to the entry point class used to start the application (see item below about `ClientFileSystemHelper`).
    * If this turns out to work well, I'll probably follow-up in a separate PR to replace the recently-added `GameRunner$Context` class with the use of services like this.  This will result in many fine-grained interfaces rather than the single coarse-grained `GameRunner$Context` type, which would seem to follow the _Interface segregation principle_.
* Modified `ClientFileSystemHelper#getCodeSourceFolder()` to use `ApplicationContext#getMainClass()` as the code source reference point rather than a hard-coded class from the `game-core` project.  This change allows this method to find the appropriate code source folder in a development environment (i.e. directory based rather than jar based) regardless of whether the game is running from the `game-headed` project folder or the `game-headless` project folder.
* Modified `ResourceLoader#getMapResourceLoader()` to make the presence of the _assets_ folder optional.  If it's not found, it won't be added to the list of locations searched for resources.
    * Even if, for some reason, a headed client was missing its _assets_ folder, a request to load a resource that would normally be found in that folder will still fail in downstream code.

## Refactoring Changes

* Moved the `HeadedGameRunner` and `HeadlessGameRunner` classes to more appropriate packages.

## Manual Testing Performed

* Verified I could run the headless game server in the following configurations:
    * Eclipse launch configuration
    * Gradle `:game-headless:run` task
    * Portable build
* Verified I could run the headed game client in the following configurations:
    * Eclipse launch configuration
    * Gradle `:game-headed:run` task
    * Portable build
    * Unix installer
    * Windows x86_64 installer
* Tested connecting one of the above clients to one of the above servers to verify I could play the following maps:
    * Classic
    * TWW
    * WAW
* Verified I could join a bot via the lobby and start a Classic game.
* Verified the PBEM/PBF dice server list was populated correctly when running the headed game client in the following configurations:
    * Eclipse launch configuration
    * Gradle `:game-headed:run` task
    * Portable build
    * Unix installer
    * Windows x86_64 installer

## Additional Review Notes

* I think I've tested things thoroughly, but something is still niggling me that I may have missed some hidden usage of the _assets_ folder within a headless game server.  @DanVanAtta, @RoiEXLab, @ron-murhammer: are you aware of any other use of the _assets_ folder that would prevent a bot from running (or just raising exceptions in the log) if it were missing?
* If anyone is feeling adventurous, build artifacts for this PR are available [here](https://github.com/ssoloff/triplea-game-triplea/releases/tag/ssoloff-1.9.0.0.2869).  It would be interesting to hear if, at least for the headed game client, there are no issues running from the portable build or the installer on a machine other than my own.
